task/monorail-server/src/ignition.rs

```diff
@@ -0,0 +1,84 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use drv_ignition_api::{Ignition, IgnitionError};
+use ringbuf::*;
+use userlib::*;
+use vsc7448::{config::PortMap, Vsc7448, Vsc7448Rw, VscError};
+
+#[derive(Copy, Clone, PartialEq)]
+enum Trace {
+    None,
+    DisableVsc7448Port(u8),
+    EnableVsc7448Port(u8),
+    VscError(VscError),
+    IgnitionError(IgnitionError),
+    Presence(u64),
+}
+ringbuf!(Trace, 48, Trace::None);
+
+task_slot!(IGNITION, ignition);
+
+pub struct IgnitionWatcher {
+    ignition: Ignition,
+
+    // There are 35 ignition channels, corresponding to
+    // - 32 sleds
+    // - 2 PSCs
+    // - 1 remote Sidecar
+    //
+    // (the local Sidecar is not available on its own Ignition network)
+    //
+    // When we boot, the VSC7448 code brings every single port up; we then
+    // disable them depending on Ignition presence detect bits.
+    enabled: [bool; 35],
+}
+
+/// Mapping from Ignition presence bit to VSC7448 port
+///
+/// This is hard-coded and the same for Sidecar rev A and B
+const IGNITION_TO_VSC7448: [u8; 35] = [
+    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
+    21, 51, 52, 24, 25, 26, 27, 28, 29, 30, 31, 40, 41, 42,
+];
+
+impl IgnitionWatcher {
+    pub fn new() -> Self {
+        let t: TaskId = IGNITION.get_task_id();
+        let ignition = Ignition::from(t);
+        Self {
+            ignition,
+            enabled: [true; 35],
+        }
+    }
+
+    pub fn wake<R: Vsc7448Rw>(&mut self, vsc7448: &Vsc7448<R>, map: &PortMap) {
+        let presence = match self.ignition.presence_summary() {
```

Comment:
To Robert's point - it seems like there are a couple ways this could go wrong:

1. If we can't get the presence summary, we won't change whether any ports are enabled/disabled. If we previously disabled all ports that didn't have a sled present, then we start failing to get the presence summary, then we plug in a new sled while we still can't get presence, it won't get its port turned on.
2. If we succeed in getting the presence summary but it incorrectly says a sled is not present on some port when it actually is, we're shutting off management network access to that sled. (Enabling a port if the presence summary incorrectly says a sled is present when it's not seems relatively harmless.)

I have no intution for how likely it is for ignition to fail (in either of these ways) while everything else is still okay, but it seems plausible?